lib/Support/Pipeline.cpp

```diff
@@ -93,29 +93,101 @@ void MappingTraits<offloadtest::DescriptorSet>::mapping(
+      // single buffer output
+      llvm::MutableArrayRef<T> Arr(reinterpret_cast<T *>(B.Data.back().get()),
+                                   B.Size / sizeof(T));
+      I.mapRequired("Data", Arr);
+    } else {
+      // array of buffers output
+      llvm::SmallVector<llvm::MutableArrayRef<T>> Arrays;
+      for (const auto &D : B.Data)
+        Arrays.emplace_back(reinterpret_cast<T *>(D.get()), B.Size / sizeof(T));
+      I.mapRequired("Data", Arrays);
+    }
+    return;
+  }
+
+  // zero-initialized buffer(s)
+  int64_t ZeroInitSize;
+  I.mapOptional("ZeroInitSize", ZeroInitSize, 0);
+  if (ZeroInitSize > 0) {
+    B.Size = ZeroInitSize;
+    for (uint32_t i = 0; i < B.ArraySize; i++) {
+      B.Data.push_back(std::make_unique<char[]>(B.Size));
+      memset(B.Data.back().get(), 0, B.Size);
+    }
+    return;
+  }
+
+  // single buffer input
+  if (B.ArraySize == 1) {
+    llvm::SmallVector<T, 64> Arr;
     I.mapRequired("Data", Arr);
-  } else {
-    int64_t ZeroInitSize;
-    I.mapOptional("ZeroInitSize", ZeroInitSize, 0);
-    if (ZeroInitSize > 0) {
-      B.Size = ZeroInitSize;
-      B.Data.reset(new char[B.Size]);
-      memset(B.Data.get(), 0, B.Size);
-      I.mapOptional("OutputProps", B.OutputProps);
+    B.Size = Arr.size() * sizeof(T);
+    B.Data.push_back(std::make_unique<char[]>(B.Size));
+    memcpy(B.Data.back().get(), Arr.data(), B.Size);
+    return;
+  }
+
+  // array of buffers input
+  llvm::SmallVector<llvm::SmallVector<T>> Arrays;
+  I.mapRequired("Data", Arrays);
+  B.Size = Arrays.back().size() * sizeof(T);
+
+  uint32_t ActualSize = 0;
+  for (auto Arr : Arrays) {
+    if (Arr.size() * sizeof(T) != B.Size) {
+      I.setError("All buffers must have the same size.");
       return;
     }
-    llvm::SmallVector<Type, 64> Arr;
-    I.mapRequired("Data", Arr);
-    B.Size = Arr.size() * sizeof(Type);
-    B.Data.reset(new char[B.Size]);
-    memcpy(B.Data.get(), Arr.data(), B.Size);
+    B.Data.push_back(std::make_unique<char[]>(B.Size));
+    memcpy(B.Data.back().get(), Arr.data(), B.Size);
+    ActualSize++;
   }
+  if (ActualSize != B.ArraySize)
+    I.setError(Twine("Expected ") + std::to_string(B.ArraySize) +
+               " buffers, found " + std::to_string(ActualSize));
+}
 
-  I.mapOptional("OutputProps", B.OutputProps);
+// Counter(s) can contain one counter value for a singular resource
+// or multiple values for array of resources with counters.
+// For a single resource the yaml will look like this:
+//
+//   Counter: 5
+//
+// For an array of resources the yaml will have look like this:
+//
+//   Counters: [5, 4, 3]
+//
+static void setCounters(IO &I, offloadtest::Buffer &B) {
+  // counters are printed only if they exist and only on output
+  if (!I.outputting() || B.Counters.empty())
+    return;
+
+  if (B.ArraySize == 1) {
+    assert(B.Counters.size() == 1 &&
+           "expected a single counter for a single buffer");
+    I.mapRequired("Counter", B.Counters.back());
```

Comment:
Can't this just be
`I.mapRequired("Counter", B.Counters);` ?

Reply:
That would output `Counter: [ 1 ]` even if the resource is not an array. I prefer to keep it as `Counter: 1`.